### DISCUSSION TITLE
[hyper-1.0] Why `http2::handshake` has an executor parameter, while `http1::handshake` doesn't?

### DISCUSSION THREAD

**Discussion by slinkydeveloper** [hyperium/hyper#3181] [Q&A]
**Version**
Hyper 1.0-rc3

**Description**
I've started looking into porting my application to hyper 1.0, and I'm currently going through the hyper 1.0 APIs to understand all the various details. One thing got my attention: why `http2::handshake` gets an executor parameter, while `http1::handshake` doesn't? I would expect the `Connection` I get out of `handshake` to be the only task spawned for the tcp connection, handling the http/2 state, but apparently it's not. Following the code I see that this executor is sometimes invoked for implementing the ping pong, right? Or am I missing something else? 

Anyway, thank you so much for Hyper 1.0!

**Comment by seanmonstar** [comment]
Are you asking about the client side, or the server side? The server also needs an executor to handle all concurrent requests on the same connection. This is because multiple requests can come in while a response isn't ready, and so those get treated by new tasks.

**Reply by slinkydeveloper** [reply]
I'm sorry, i forgot to say it's client side :)

**Comment by ncihnegn** [comment]
Why do clients need it?
